lib/galaxy/util/__init__.py

```diff
@@ -1627,6 +1627,21 @@ def url_get(base_url, auth=None, pathspec=None, params=None, max_retries=5, back
     return response.text
 
 
+def is_url(uri, allow_list=None):
+    """
+    check if uri is (most likely) an url, more precisely the function checks
+    if the uri starts with a protocol from the allow list (defaults to
+    "http://", "https://", "ftp://")
+    >>> is_url('https://zenodo.org/record/4104428/files/UCSC-hg38-chr22-Coding-Exons.bed')
+    True
+    >>> is_url('/some/path')
+    False
+    """
+    if allow_list is None:
+        allow_list = ["http://", "https://", "ftp://"]
+    return any([uri.startswith(_) for _ in allow_list])
```

Comment:
That's not exactly the same as `str.find`, are we sure that wasn't used intentionally (i.e to deal with leading spaces) ?

Reply:
We can strip spaces inside this function, but then the caller may pass the URL to other functions that don't accept leading spaces. It's still actually quite a lenient checker (we could use e.g. `urllib.parse.urlparse()` for stricter validation).

Reply:
I don't disagree with that, however that is not how we used it in upload.py, and we should at least do some manual checks there in the absence of tests.

Reply:
> we should at least do some manual checks there in the absence of tests.

can do, but is there no test at all?

Reply:
It's somewhat of a odd corner case, right ? I think the intention is to actually strip leading spaces early on (https://github.com/mvdbeek/galaxy/blob/e7c827727e165cfe4a6e24c1f07dc1796772d980/lib/galaxy/tools/actions/upload_common.py#L35) so maybe the startswith variant is fine. AFAICT we have no tests asserting the behavior (at least I didn't find anything I could find by grepping for `leading`).

Reply:
There should be tests for https://github.com/galaxyproject/galaxy/blob/dev/tools/data_source/upload.xml that we can extend, or? I guess this one would be easy to extend to contain leading spaces https://github.com/galaxyproject/galaxy/blob/e7c827727e165cfe4a6e24c1f07dc1796772d980/lib/galaxy_test/api/test_tools_upload.py#L776

@nsoranzo regarding `file://` I found https://github.com/galaxyproject/galaxy/blob/e7c827727e165cfe4a6e24c1f07dc1796772d980/test/integration/test_upload_configuration_options.py#L11

Reply:
Does someone know about `file://`?

Reply:
`file://` is handled explicitly upstream ... I think it should be fine as at is now, but see my comment below.